src/modules/job-exec/exec.c

```diff
@@ -107,17 +120,87 @@ static void complete_cb (struct bulk_exec *exec, void *arg)
                             bulk_exec_rc (exec));
 }
 
-static int exec_barrier_enter (struct bulk_exec *exec)
+static void start_timer_cb (flux_reactor_t *r,
+                            flux_watcher_t *w,
+                            int revents,
+                            void *arg)
+{
+    struct exec_ctx *ctx = arg;
+    struct jobinfo *job = ctx->job;
+    struct bulk_exec *exec = ctx->job->data;
+    char *ranks;
+
+    if (!(ranks = idset_encode (ctx->barrier_pending_ranks,
+                                IDSET_FLAG_RANGE))) {
+        flux_log_error (job->h,
+                        "failed to encode barrier pending ranks for job %s",
+                        idf58 (job->id));
+        return;
+    }
+
+    (void) jobinfo_drain_ranks (job,
+                                ranks,
+                                "job %s start timeout: %s",
+                                idf58 (job->id),
+                                "possible node hang");
+
+    jobinfo_fatal_error (job,
+                         0,
+                         "%s waiting for %d/%d nodes (rank%s %s)",
+                         "start barrier timeout",
+                         idset_count (ctx->barrier_pending_ranks),
+                         bulk_exec_total (exec),
+                         idset_count (ctx->barrier_pending_ranks) > 1 ?"s":"",
+                         ranks);
+    free (ranks);
+}
+
+static void barrier_timer_stop (struct exec_ctx *ctx)
+{
+    if (ctx->shell_barrier_timer)
+        flux_watcher_stop (ctx->shell_barrier_timer);
+}
+
+static void barrier_timer_start (struct exec_ctx *ctx)
+{
+    flux_reactor_t *r = flux_get_reactor (ctx->job->h);
+    double timeout = ctx->barrier_timeout;
+
+    /* Only ever create one barrier timer (for the first shell barrier)
+     */
+    if (ctx->shell_barrier_timer && timeout > 0.)
+        return;
+
+    /* Start a timer for all job shells to reach the shell barrier.
+     * If the timer expires, then an exception will be raised.
+     */
+    if (!(ctx->shell_barrier_timer = flux_timer_watcher_create (r,
+                                                                timeout,
+                                                                0.,
+                                                                start_timer_cb,
+                                                                ctx)))
+        flux_log_error (ctx->job->h,
+                        "%s: failed to create barrier start timer",
+                        idf58 (ctx->job->id));
+    else
+        flux_watcher_start (ctx->shell_barrier_timer);
```

Comment:
It might be better to create the watcher when the exec_ctx is created and just not start it, so this error handling wouldn't be required in a place where there's nothing to do but log it.

Reply:
Good suggestion, though this breaks detection of the first barrier elsewhere (also not sure it should be _fatal_ if we can't start the timer anyway, in the common case the job will still proceed normally).

I think there's another variable I can use to detect the first barrier so let me see if I can rework this.

Reply:
In case I wasn't clear I meant create the watcher up front so there would be no error handling in this function.  Then all the creation functions with unlikely failure (e.g. ENOMEM) can be grouped together with a common error path.

Reply:
Yeah, that was clear. My point was that currently the code uses `timer_watcher != NULL` to indicate other barriers than the first, and skips starting the timer in these cases.

Also, that maybe failure to create the barrier timer should not be fatal to the job as implied. However, as you say, the failure is unlikely (so I guess that's why I thought it was ok to do it outside of the constructor), so I'll attempt to move it. It has a cascading effect though so this will take awhile.

Reply:
Eh if you feel it's better the way it is, I'm fine with that.  It was just a suggestion.

Reply:
No, I think you're right that all the constructors should occur up front, if only just so they are easy to find. The watcher will be created unnecessarily for single-node jobs, but that is not a big deal.